l10n_ch: define a specific grouping rule for ISR payments

yvaucher opened the pull request:
Swiss ISR payments must not be grouped as we want to keep a single transaction
per reference. The ISR payment reference doesn't imply additional fees from
the bank.

One transaction per ISR reference is needed in Swiss SEPA payments
to ensure a end-to-end flow that will ease the reconciliation on
the other end.

When ISR reference is detected we also avoid to concatenate the references
to not concatenate the same reference multiple times. In case the
same reference is sent by the supplier on different dates.

**Description of the issue/feature this PR addresses:**

This is about SEPA payment in `account_sepa` module and the source of the issue is the wizard 'Register payment'. On that wizard, There is an option 'Group invoices' that you can check or not.

In Switzerland, we group all IBAN payments (to limit bank fees as everywhere in europe) but we never group ISR payments (because there are no fees and the supplier can't reconcile automatically if we group them - then they are not happy)

**Current behavior before PR:**

Payments are made difficult because to get the good grouping you would have to check each invoice one by one to know which you can group together or not. And you must even go at bank account detail of the partner to check if it is an ISR payment or not.

This give 3 options in the current state:

1. You loose lots of hours and risks a lot of mistake by having to check invoice and partner bank one by one (not really viable)
2. You group everything and your suppliers will have trouble to reconcile.
3. You group nothing and it costs you a lot of fees that you could have avoided.

**Desired behavior after PR is merged:**

Being able to group everything but with a special rules for structured ISR references.

This was discussed in OPW #2 176 540 following solutions made for Swiss SEPA payments in enterprise module `account_sepa`.


--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

yvaucher commented:
@oco-odoo @fclementic2c

fclementic2c commented:
@yvaucher 

I did a test : https://drive.google.com/open?id=1yasmxyRZkkFaJGsY_8pXsY-IjQhXPbc5
but I have an error, can you check it please?

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/data/build/odoo/odoo/http.py", line 619, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/data/build/odoo/odoo/http.py", line 309, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/data/build/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/data/build/odoo/odoo/http.py", line 664, in dispatch
    result = self._call_function(**self.params)
  File "/data/build/odoo/odoo/http.py", line 345, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/data/build/odoo/odoo/service/model.py", line 93, in wrapper
    return f(dbname, *args, **kwargs)
  File "/data/build/odoo/odoo/http.py", line 338, in checked_call
    result = self.endpoint(*a, **kw)
  File "/data/build/odoo/odoo/http.py", line 910, in __call__
    return self.method(*args, **kw)
  File "/data/build/odoo/odoo/http.py", line 510, in response_wrap
    response = f(*args, **kw)
  File "/data/build/odoo/addons/web/controllers/main.py", line 1324, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/data/build/odoo/addons/web/controllers/main.py", line 1312, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 395, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 382, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/data/build/odoo/addons/account/models/account_payment.py", line 848, in create_payments
    payments = Payment.create(self.get_payments_vals())
  File "/data/build/odoo/addons/account/models/account_payment.py", line 834, in get_payments_vals
    grouped[self._get_payment_group(inv)] += inv
  File "/data/build/odoo/addons/l10n_ch/models/account_payment.py", line 27, in _get_payment_group
    if inv._is_isr_supplier_invoice():
  File "/data/build/odoo/addons/l10n_ch/models/account_invoice.py", line 224, in _is_isr_supplier_invoice
    re.match(r'^(\d{2,27}|\d{2}( \d{5}){5})$', self.ref)
  File "/usr/lib/python3.6/re.py", line 172, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object

yvaucher commented:
@fclementic2c thanks I assumed wrongly that ref would always contains a string. But an empty ref is `False`.

I fixed the issue and added a test on this.

fclementic2c commented:
@yvaucher I just did a grouped payment test here : http://1494262-45744-654dc0.runbot36.odoo.com/web?reload=#id=7&active_id=2&model=account.payment&view_type=form&cids=3&menu_id=384 

... but the behaviour did not change: Odoo has created account.payment grouped by supplier as usual.
Can you check please ?

yvaucher commented:
@fclementic2c ok, my mistake was to use `ref` instead of `invoice_payment_ref`, I rewrote the first commit to use `invoice_payment_ref` plus I added a second commit that still allows to use `ref`. Which according to the copy of the value from invoice to communication makes sense and will be more user error prone.

oco-odoo commented on addons/l10n_ch/models/account_payment.py:
```diff
@@ -0,0 +1,31 @@
+        we want a unique reference in communication.
+
+        """
+        if invoices[0]._is_isr_supplier_invoice():
```
This is a little weird, since you're actually depending on the function below to do that... At least, put some comment explaining that.

yvaucher replied:
Maybe the docstring of the method is not enough? How do you want me to improve it?

Maybe I should add an example?

In short we want that:

`120000000000234478943216899 120000000000234478943216899`
be
`120000000000234478943216899`

This is tested in the unit tests.

yvaucher replied:
So this is only for communications. As shown in the example, we want to avoid having a repetition of the same structured ref that would break the generation of SEPA file with structured ref.

yvaucher replied:
I added a comment in the code.

yvaucher commented on addons/account/models/account_payment.py:
```diff
@@ -810,17 +817,21 @@ def _prepare_payment_vals(self, invoices):
-                grouped[(inv.commercial_partner_id, inv.currency_id, inv.invoice_partner_bank_id, MAP_INVOICE_TYPE_PARTNER_TYPE[inv.type])] += inv
-            else:
-                grouped[inv.id] += inv
+            grouped[self._get_payment_group(inv)] += inv
```
I consider `inv.id` as a default grouping. Even if no grouping is defined we are using the `grouped` variable.
This allow for a better hook. Moving the if to `_get_payment_group_key` means having the logic for grouping key on 2 different places.

yvaucher replied:
```suggestion
            grouped[self._get_payment_group_key(inv)] += inv
```

yvaucher commented on addons/account/models/account_payment.py:
```diff
@@ -810,17 +817,21 @@ def _prepare_payment_vals(self, invoices):
         }
         return values
 
+    def _get_payment_group(self, inv):
```
Sounds better indead.
```suggestion
    def _get_payment_group_key(self, inv):
```

yvaucher commented on addons/l10n_ch/models/account_payment.py:
```diff
@@ -0,0 +1,31 @@
+        else:
+            return super()._prepare_communication(invoices)
+
+    def _get_payment_group(self, inv):
```
```suggestion
    def _get_payment_group_key(self, inv):
```

yvaucher commented on addons/l10n_ch/models/account_payment.py:
```diff
@@ -0,0 +1,31 @@
+            ref = inv.invoice_payment_ref or inv.ref
+            return (inv.commercial_partner_id, inv.currency_id, inv.invoice_partner_bank_id, ref)
+        else:
+            return super()._get_payment_group(inv)
```
```suggestion
            return super()._get_payment_group_key(inv)
```

yvaucher commented on addons/l10n_ch/models/account_payment.py:
```diff
@@ -0,0 +1,31 @@
+        As such reference is structured. This is required to export payments
+        to bank in batch.
+        """
+        if inv._is_isr_supplier_invoice():
```
Ok, makes sense, in a way we always want them ungrouped.
But you are right we could have in rare case twice the same reference and in this case grouping or not is a choice you have.

fclementic2c commented:
New functional test.
Works great !

yvaucher commented:
@oco-odoo I moved back the `if self.group_payment`.

I kept separated commits to ease the review, to squash later.

yvaucher commented:
Rebased to fix conflict on `addons/l10n_ch/models/__init__.py`

yvaucher commented:
@oco-odoo what if you only accept the `_get_payment_group_key` and `_get_communication` hooks part to let use change the behavior easily from an external module. Could it be accepted in Odoo 13.0 ?

yvaucher commented:
ping @oco-odoo

oco-odoo commented:
@yvaucher If it's only for two hooks,no problem, we can do that in stable :)

yvaucher commented:
@oco-odoo ok, I'll change this when time permits it.

yvaucher commented:
@oco-odoo does https://github.com/odoo/odoo/pull/48441 looks good to you (in version 12.0)
I can do the same for version 13.0 if needed.

oco-odoo commented:
@yvaucher I just r+ed https://github.com/odoo/odoo/pull/48441 , so I'm closing this PR ;)

yvaucher commented:
@oco-odoo Thanks for the version 12.0 there might be some conflicts to integrate it in 13.0. Please tell me if you need assistance.
